src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -92,162 +98,346 @@ fn default_endpoint() -> String {
+                        .and_modify(|cur| cur.merge(&data_point))
+                        .or_insert(data_point);
+
+                    if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                        break;
+                    }
+                }
+            }
+
+            // We send a timeseries create request as long as we have at least one metric data point and
+            // we waited enough since the last time we sent something to GCP.
+            if !aggregate.is_empty()
+                && last_time.elapsed() >= GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY
+            {
+                let start_time = time_tracker.start_time();
+                let time_series = aggregate
+                    .drain()
+                    .map(|t| gcp_timeseries(&self.config, start_time, t.1))
+                    .collect::<Vec<_>>();
+
+                let req = googleapis_proto::CreateTimeSeriesRequest {
+                    name: format!("projects/{}", self.config.project_id),
+                    time_series,
                 };
 
-                (*value, interval, gcp::GcpMetricKind::Cumulative)
+                let mut attempts = 1usize;
+
+                loop {
+                    if let Err(status) = client.create_time_series(Request::new(req.clone())).await
+                    {
+                        if status.code() == Code::Internal
+                            || status.code() == Code::Unknown
+                                && attempts < MAX_RETRIES_ON_LEGITIMATE_GCP_ERROR
+                        {
+                            tokio::time::sleep(Duration::from_millis(500)).await;
+                            attempts += 1;
+                            continue;
+                        }
+
+                        error!(message = "Error when sending time_series", %status);
+                        emit!(ComponentEventsDropped::<UNINTENTIONAL> {
+                            count: attempts,
+                            reason: "Error when sending timeseries to GCP",
+                        });
+                    }
+
+                    break;
+                }
+
+                last_time = Instant::now();
             }
-            MetricValue::Gauge { value } => {
-                let interval = gcp::GcpInterval {
-                    start_time: None,
-                    end_time,
-                };
 
-                (*value, interval, gcp::GcpMetricKind::Gauge)
+            tokio::select! {
+                _ = stopwatch.tick() => {}
+
+                events = input.next() => {
+                    if events.is_none() {
+                        return Ok(());
+                    }
+
+                    if let EventArray::Metrics(metrics) = events.unwrap() {
+                        let start_time = time_tracker.start_time();
+                        buffer.extend(metrics.into_iter().filter_map(|m| MetricDataPoint::from_metric(&self.config, start_time, m)));
+                    }
```

Comment:
There's a new `let else` structure that you can use here:
```suggestion
                    let Some(events) = events else { return Ok(()) };

                    if let EventArray::Metrics(metrics) = events {
                        let start_time = time_tracker.start_time();
                        buffer.extend(metrics.into_iter().filter_map(|m| MetricDataPoint::from_metric(&self.config, start_time, m)));
                    }
```